rxjs-1/leaderBoard.js

```diff
@@ -0,0 +1,65 @@
+const { fromEvent, from } = require('rxjs');
+const { reduce, map, throttleTime, flatMap, bufferCount } = require('rxjs/operators');
+const Table = require('easy-table');
+
+const { getRace } = require('./race');
+const { round } = require('./util');
+
+const race = getRace();
+const carNames = race.getCars();
+
+console.log('Participants:', carNames.join('/'));
+
+function sortByPosition(cars) {
+  return cars.sort((x, y) => y.xLocation - x.xLocation)
+}
+
+function raceReducer(acc, curr, index) {
+  const [leader, ...others] = acc
+  const isLast = index + 1 === carNames.length
+  return [
+    // reset first object once all gaps are calculated
+    ...(isLast ? [{ ...leader, leaderGapDistance: 0, leaderGapTime: 0 }, ...others] : acc),
+    {
+      position: index + 1,
+      carName: curr.carName,
+      leaderGapDistance: leader
+        ? round(leader.leaderGapDistance - curr.xLocation)
+        : curr.xLocation,
+      leaderGapTime: leader
+        ? round(leader.leaderGapTime - curr.time)
+        : curr.time,
+    },
+  ]
+}
+
+function getLeaderBoard(race) {
+  return fromEvent(race, 'data').pipe(
+    bufferCount(carNames.length),
```

Comment:
Here you assume that you will always receive the cars is a recurring and predictable manner, i.e `A0 -> B0 -> A1 -> B1 -> A2 -> B2` ? Otherwise you would get a buffer with the same car twice, but at different time.

Reply:
Yes I'm assuming to receive the cars by pair, no matter the time nor the order cos I sort the list just after. It wouldn't work with a different sequence like `A0 -> A1 -> B0 -> B1 -> A2 -> A3`, but I choose `bufferCount` as a shortcut because I saw how race event is built.

A solution to handle different sequences could be treating each car race separately (filtering by carName) then combining the latest emission of them in `combineLatest`.